src/oic/oic/provider.py

```diff
@@ -1594,14 +1592,11 @@ def read_registration(self, authn, request, **kwargs):
             return error_response('invalid_request')
         token = authn[len("Bearer "):]
 
-        try:
-            client_id = self.cdb[token]
-        except KeyError:
-            return Unauthorized()
-
-        # extra check
+        # Get client_id from request
         _info = parse_qs(request)
-        if not _info["client_id"][0] == client_id:
+        client_id = _info.get("client_id", [None])[0]
+
+        if self.cdb.get(client_id) is None or self.cdb.get(client_id)['registration_access_token'] != token:
```

Comment:
The access token comparision should use a constant time compare function, to avoid a timing attack.
It should also ALWAYS check the access token against a dummy value, if the client_id does not exist.

    from jwkest import safe_str_cmp
    cdb_entry = self.cdb.get(client_id)
    reg_token = cdb_entry['registration_acccess_token'] if cdb_entry else ''
    token_broken = not safe_str_cmp(reg_token, token)
    if cdb_entry is None or token_broken:
        ...

Otherwise someone could do a timing attack to guess the registration_access_token and get access to the client_secret. This wasn't possible with the old style, as that did a hash on the token, which basically is constant time and looked it up in the cdb.